#!/bin/bash
#this script is used to sort/map mutiple sequences downloaded from public database with reference strain NCTC11168. 
#run this script after activating related conda environment: in my research, is called "sequence" 
#used tools: 
#	bwa: 		for alignment process
#	samtools sort: 	sort the sequences in the order of alignement
#	samtools fasta:	write the .bam filed generated by bwa to a new fasta file. Here .fna is used instead of .fasta
#	GNU parallel:	tools to use mutiple cpu core at the same time, increase the proccess speed.

# make reference strain index
bwa index /home/vader/Documents/MSc-project/data/processed_data/local_library/nctc11168_mapping_library/GCF_000009085.1_ASM908v1_genomic.fna


data_path="/home/vader/Documents/MSc-project/data/processed_data/RefSeq_available_for_analysis"

# Change to the data folder
cd "$data_path"

# Define a function to process a single subfolder
process_folder() {
    dir="$1"
    if [ -d "$dir" ]; then
    	
    	dir_name=$(basename "$dir")

        # Loop for GCF file in the current subfolder
        for file in "$dir"/GCF_*.fna; do
        
            # Run samtools using absolute paths
            bwa mem -p "/home/vader/Documents/MSc-project/data/processed_data/local_library/nctc11168_mapping_library/GCF_000009085.1_ASM908v1_genomic.fna" "$file" | samtools sort -O bam -l 0 | samtools fasta -0 "$dir/mapped_genome.fna" -
            
            sed -e '/^>/d' "$dir/mapped_genome.fna" > "$dir/mapped_genome.tmp1"
            echo "> $dir_name" | cat - "$dir/mapped_genome.tmp1" > "$dir/mapped_genome.tmp2"
            fold -w 80 "$dir/mapped_genome.tmp2" > "$dir/mapped_genome.fna"
            rm "$dir/mapped_genome.tmp1" "$dir/mapped_genome.tmp2"
            
        done
    fi
}

export -f process_folder

find . -mindepth 1 -maxdepth 1 -type d -print0 | parallel -j 80% -0 process_folder 

